pkg/imageacquirer/docker-image-acquirer.go

```diff
@@ -0,0 +1,263 @@
+package imageacquirer
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"time"
+
+	docker "github.com/fsouza/go-dockerclient"
+	iiapi "github.com/openshift/image-inspector/pkg/api"
+	"github.com/openshift/image-inspector/pkg/util"
+)
+
+const (
+	PULL_LOG_INTERVAL_SEC = 10 * time.Second
+)
+
+type dockerImageAcquirer struct {
+	DockerSocket        string       // DockerSocket for the local docker daemon
+	PreferedDestination string       // PreferedDestination for the extraction of the image
+	PullPolicy          string       // PullPolicy controls whether we try to pull the inspected image
+	Auths               AuthsOptions // Sources for authentications with docker
+}
+
+func (dii *dockerImageAcquirer) Acquire(source string) (string, docker.Image, iiapi.ScanResult, iiapi.FilesFilter, error) {
+	client, err := docker.NewClient(dii.DockerSocket)
+	if err != nil {
+		return "", docker.Image{}, iiapi.ScanResult{}, nil,
+			fmt.Errorf("Unable to connect to docker daemon: %v\n", err)
+	}
+	imageMetaBefore, inspectErrBefore := client.InspectImage(source)
+	if dii.PullPolicy == iiapi.PullNever && inspectErrBefore != nil {
+		return "", docker.Image{}, iiapi.ScanResult{}, nil,
+			fmt.Errorf("Image %s is not available and pull-policy %s doesn't allow pulling",
+				source, dii.PullPolicy)
+	}
+
+	if dii.PullPolicy == iiapi.PullAlways ||
+		(dii.PullPolicy == iiapi.PullIfNotPresent && inspectErrBefore != nil) {
+		if err = dii.dockerPullImage(source, client); err != nil {
+			return "", docker.Image{}, iiapi.ScanResult{}, nil, err
+		}
+	}
+
+	imageMetaAfter, inspectErrAfter := client.InspectImage(source)
+	if inspectErrBefore == nil && inspectErrAfter == nil &&
+		imageMetaBefore.ID == imageMetaAfter.ID {
+		log.Printf("Image %s was already available", source)
+	}
+
+	imageMetadata, dstPath, err := dii.extractImageFromContainer(client, source)
+	if err != nil {
+		return "", docker.Image{}, iiapi.ScanResult{}, nil, err
+	}
+
+	scanResults := iiapi.ScanResult{
+		APIVersion: iiapi.DefaultResultsAPIVersion,
+		ImageName:  source,
+		ImageID:    imageMetadata.ID,
+		Results:    []iiapi.Result{},
+	}
+	return dstPath, *imageMetadata, scanResults, nil, nil
+}
+
+// dockerPullImage pulls the inspected image through the docker socket
+// using the given client.
+// It will try to use all detected authentication methods and will fail
+// only if all of them failed.
+func (dii *dockerImageAcquirer) dockerPullImage(source string, client *docker.Client) error {
+	log.Printf("Pulling image %s", source)
+
+	var imagePullAuths *docker.AuthConfigurations
+	var authCfgErr error
+	if imagePullAuths, authCfgErr = getAuthConfigs(dii.Auths); authCfgErr != nil {
+		return authCfgErr
+	}
+
+	// Try all the possible auth's from the config file
+	var err error
+	for name, auth := range imagePullAuths.Configs {
+		parsedErrors := make(chan error, 100)
+		defer func() { close(parsedErrors) }()
+
+		go func() {
```

Comment:
@enoodle why do we put this block under a goroutine and then create another goroutine to decode the response?
Seem's to me like `parsedErrors` blocks anyways until `client.PullImage` and then `parsedError` blocks for `parsedErrors` so essentially the bottleneck is on `client.PullImage` So the following lines (the ones outside the goroutine) won't be running concurrently anyhow but after the goroutine is done.

Reply:
@nimrodshn This code was moved from `pkg/inspector/image-inspector.go` and is not new here,
If I remember correctly this is done due to the error handling/writing to pipes and closing the other goroutines and channels. It is not and issue of bottle neck.
original PR: https://github.com/openshift/image-inspector/pull/42

Reply:
Yeah I just looked it over and it looked wiered to me - Reading the comments in the PR it makes sense now 😃  👍